File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
	"math/rand"
	"math"
	"sort"
)

var BASE_HITPOINTS int = 10

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

type Dice []int

func (d Dice) Len() int {
	return len(d)
}

func (d Dice) Swap(i, j int) {
	d[i], d[j] = d[j], d[i]
}

func (d Dice) Less(i, j int) bool {
	return d[i] < d[j]
}

func (d Dice) Sum() int {
	sum := 0
	for _, val := range d {
		sum += val
	}

	return sum
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor(float64(score-10) / 2.0))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	dice := Dice{}
	for i:=0; i<4; i++ {
		dice = append(dice, randInt(1, 7))
	}

	sort.Sort(dice)

	return dice[1:].Sum()
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	constitution := Ability()
	hitpoints := Modifier(constitution) + BASE_HITPOINTS

	return Character{
		Strength: Ability(),
		Dexterity: Ability(),
		Constitution: constitution,
		Intelligence: Ability(),
		Wisdom: Ability(),
		Charisma: Ability(),
		Hitpoints: hitpoints,
	}
}

func randInt(min, max int) int {
	return min + rand.Intn(max - min)
}

File: solutions/ruby/dnd-character/1/dnd_character.rb
module Die
  def self.roll(roll: 4, die: 6, remove_lowest: 1)
    roll.times.map { rand(1..die) }.sort.pop(roll - remove_lowest).sum
  end
end

class DndCharacter
  BASE_HITPOINTS = 10
  private_constant :BASE_HITPOINTS

  attr_reader :strength,
    :dexterity,
    :constitution,
    :intelligence,
    :wisdom,
    :charisma,
    :hitpoints

  def initialize
    @strength     = Die.roll
    @dexterity    = Die.roll
    @constitution = Die.roll
    @intelligence = Die.roll
    @wisdom       = Die.roll
    @charisma     = Die.roll

    @hitpoints    = BASE_HITPOINTS + self.class.modifier(constitution)
  end

  def self.modifier(ability_score)
    ability_score / 2 - 5
  end
end